### ISSUE TITLE
Windows: Get all unique API IDs on device input

### ISSUE THREAD

**Issue by unexpectedpanda** [pyglet/pyglet#1137]
I'm trying to use Pyglet to connect the unique IDs from multiple input APIs back to the one hardware device in Windows. The ultimate goal is to be able to provide input from a device (mouse, keyboard, joystick, controller, whatever), and have Pyglet tell me what the device's DirectInput ID, XInput ID, and HID path are (assuming they exist) and what was pressed/moved.

After going through other issues I've read that XInput and DirectInput can't be initialized at the same time, so perhaps the user would have to select the API of choice before providing input, but I'd still like to be able to link all these disparate pieces together to help disambiguate devices.

For example, I have the following device, that has the following IDs:

| Property | Value |
|-------------|-------------|
| Name         | Arcade Stick (Street Fighter IV FightStick TE) |
| HID device path | `HID\VID_1BAD&PID_F038&IG_00\8&5D5C014&0&0000` |
| DirectInput GUID | `6eb03f00-2da6-11ef-8001-444553540000` |
| Xinput ID | XInput Player 1 |

And at this stage I'm trying to get the output:

```py
{
  'name': 'Arcade Stick (Street Fighter IV FightStick TE)',
  'directinput_device': 'DirectInputDevice(name=Arcade Stick (Street Fighter IV FightStick TE))',
  'directinput_id': '6eb03f00-2da6-11ef-8001-444553540000',
  'xinput_device': 'XInputDevice(name=XInput0)',
  'hid_path': 'HID\VID_1BAD&PID_F038&IG_00\8&5D5C014&0&0000',
  'input': ['Button(name=a, raw_name=None)']
}
```

So far I've managed to modify the [DirectInput code](https://github.com/pyglet/pyglet/blob/master/pyglet/input/win32/directinput.py) to get the DirectInput ID and tie it back to the `DirectInputDevice` (`(DirectInputDevice(name=Arcade Stick (Street Fighter IV FightStick TE)), '78a76bc0-2f4d-11ef-8008-444553540000')`). XInput has been harder -- I've been able to iterate over all HID paths for XInput devices by grabbing `guid` from `get_xinput_guids`, but I've not been able to tie the HID paths back to the specific `XInputDevice`.

I thought maybe I could use the HID path to link the DirectInput device to the XInput device using DirectInput's [`GetProperty`](http://doc.51windows.net/Directx9_SDK/input/ref/ifaces/idirectinputdevice9/getproperty.htm) with `DIPROP_GUIDANDPATH`, but I was getting type errors on the parameters. That sort of thing is definitely above my comprehension at this stage.

Basically I've hit the limit of my knowledge, and I'd be grateful for any advice pointing me in the right direction.

**Comment by caffeinepills** [comment]
I used this function to match up which guids are xinput capable. You may have to modify this a bit but maybe it will help.
https://github.com/pyglet/pyglet/blob/e58747a0a43c8be2254f39cd657d692682d96361/pyglet/input/win32/xinput.py#L286

**Comment by caffeinepills** [comment]
Can you post your code of how GetProperty was failing?

**Comment by unexpectedpanda** [comment]
Thanks! All this is new to me, so I'm not surprised I was getting it wrong. I had the class set up, but didn't know about setting up the `data` variables, and was leaving off the `ctypes` in the `GetProperty` arguments.

Here's where I'm at so far from pillaging existing Pyglet examples and with your help, but unfortunately still no luck with `DIPROP_GUIDANDPATH`.

In the code sample I've made the (possibly incorrect) assumption that since the XInput devices and their HID paths are returned in a deterministic order, that both are probably based on device enumeration order and I can pair them with a zip function as an alternative to matching them more explicitly. 

```py
import ctypes
from ctypes.wintypes import MAX_PATH, ULONG, WCHAR
from pyglet.libs.win32 import com, dinput, _ole32 as ole32, _oleaut32 as oleaut32
from pyglet.libs.win32.types import byref, VARIANT
from pyglet.libs.win32.constants import CLSCTX_INPROC_SERVER
from pyglet.input.win32.directinput import _init_directinput, DirectInputDevice
from pyglet.input.win32.xinput import (
    get_devices as get_xinput_devices,
    IEnumWbemClassObject,
    IWbemClassObject,
    IWbemServices,
    IWbemLocator,
    BSTR,
    VT_BSTR,
    RPC_C_AUTHN_WINNT,
    RPC_C_AUTHZ_NONE,
    RPC_C_AUTHN_LEVEL_CALL,
    RPC_C_IMP_LEVEL_IMPERSONATE,
    EOAC_NONE,
    CLSID_WbemLocator,
    IID_IWbemLocator
    )

def get_devices():
    # Get the XInput devices
    try:
        xinput_hid_paths = []
        xinput_devices = []

        for device in get_xinput_devices():
            xinput_devices.append(device)

        # Get the XInput device HID paths
        var = VARIANT()
        oleaut32.VariantInit(byref(var))
        locator = IWbemLocator()
        services = IWbemServices()
        enum_devices = IEnumWbemClassObject()
        devices = (IWbemClassObject * 20)()

        ole32.CoCreateInstance(CLSID_WbemLocator, None, CLSCTX_INPROC_SERVER, IID_IWbemLocator, byref(locator))

        name_space = BSTR("\\\\.\\root\\cimv2")
        class_name = BSTR("Win32_PNPEntity")
        device_id = BSTR("DeviceID")

        # Connect to WMI
        hr = locator.ConnectServer(name_space, None, None, 0, 0, None, None, byref(services))
        if hr != 0:
            return xinput_hid_paths

        # Switch security level to IMPERSONATE.
        hr = ole32.CoSetProxyBlanket(services, RPC_C_AUTHN_WINNT, RPC_C_AUTHZ_NONE, None, RPC_C_AUTHN_LEVEL_CALL,
                                    RPC_C_IMP_LEVEL_IMPERSONATE, None, EOAC_NONE)

        if hr != 0:
            return xinput_hid_paths

        hr = services.CreateInstanceEnum(class_name, 0, None, byref(enum_devices))

        if hr != 0:
            return xinput_hid_paths

        while True:
            returned = ULONG()
            _hr = enum_devices.Next(10000, len(devices), devices, byref(returned))
            if returned.value == 0:
                break
            for i in range(returned.value):
                result = devices[i].Get(device_id, 0, byref(var), None, None)
                if result == 0:
                    if var.vt == VT_BSTR and var.bstrVal != "":
                        if (
                            'IG_' in var.bstrVal
                            and 'HID\\' in var.bstrVal):
                                xinput_hid_paths.append(var.bstrVal)

        oleaut32.VariantClear(byref(var))

        # The order of the XInput devices and HID paths seems to be deterministic, so
        # is possibly based on device enumeration order. If this is so, we can match the
        # HIDs directly with the devices.
        [print(x) for x in zip(xinput_devices, xinput_hid_paths)]

    except ImportError:
        pass

    # Get the DirectInput devices
    _i_dinput = _init_directinput()
    dinput_devices = []

    def _device_enum(device_instance, arg):
        # Construct the DirectInput GUID as reported by MAME
        final_quartet: str = ''.join(
            [format(device_instance.contents.guidInstance.Data4[0], "02x"),
             format(device_instance.contents.guidInstance.Data4[1], "02x")])

        # The suffix is always 444553540000, which in ASCII is "DEST" followed by 0000
        instance_guid: str = (
                f'{format(device_instance.contents.guidInstance.Data1, "08x")}-'
                f'{format(device_instance.contents.guidInstance.Data2, "04x")}-'
                f'{format(device_instance.contents.guidInstance.Data3, "04x")}-'
                f'{final_quartet}-444553540000'
        )

        device = dinput.IDirectInputDevice8()
        _i_dinput.CreateDevice(device_instance.contents.guidInstance, byref(device), None)
        dinput_devices.append(DirectInputDevice(None, device, device_instance.contents))

        print((DirectInputDevice(None, device, device_instance.contents), instance_guid))

        # Attempt to get the HID path of each DirectInput device
        try:
            class DIPROPGUIDANDPATH(ctypes.Structure):
                _fields_ = (
                    ('diph', dinput.DIPROPHEADER),
                    ('guidClass', com.GUID),
                    ('wszPath', WCHAR * MAX_PATH),
                )

            data = DIPROPGUIDANDPATH()
            data.diph.dwSize = ctypes.sizeof(data)
            data.diph.dwHeaderSize = ctypes.sizeof(data.diph)
            data.diph.dwObj = 0
            data.diph.dwHow = dinput.DIPH_DEVICE

            device.GetProperty(ctypes.pointer(dinput.DIPROP_GUIDANDPATH), ctypes.byref(data.diph))

        except Exception as e:
            print(f'❌ GetProperty error: {e}\n')

        return dinput.DIENUM_CONTINUE

    _i_dinput.EnumDevices(dinput.DI8DEVCLASS_ALL,
                          dinput.LPDIENUMDEVICESCALLBACK(_device_enum),
                          None,
                          dinput.DIEDFL_ATTACHEDONLY)

    return dinput_devices

get_devices()

```

Output:

```
(XInputDevice(name=XInput0), 'HID\\VID_1BAD&PID_F038&IG_00\\8&5D5C014&0&0000')
(XInputDevice(name=XInput1), 'HID\\VID_1BAD&PID_F038&IG_03\\8&1C72B844&0&0000')
(DirectInputDevice(name=Mouse), '6f1d2b60-d5a0-11cf-bfc7-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=Keyboard), '6f1d2b61-d5a0-11cf-bfc7-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=Arctis Nova Pro Wireless), 'e70cfa30-db2e-11ee-8001-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=Arcade Stick (Street Fighter IV FightStick TE)), '6eb03f00-2da6-11ef-8001-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=GMMK PRO), 'e70d2140-db2e-11ee-8002-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=Arctis Nova Pro Wireless), 'e70d4850-db2e-11ee-8005-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=Logitech StreamCam), '4e574b20-fa77-11ee-8001-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=GMMK PRO), 'e70d6f60-db2e-11ee-8006-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=USB Audio), 'e70dbd80-db2e-11ee-8010-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=GMMK PRO), 'e70d6f60-db2e-11ee-8007-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=Arctis Nova Pro Wireless), 'e70d6f60-db2e-11ee-8008-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=Gaming Mouse G502), 'e70d2140-db2e-11ee-8003-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=Gaming Mouse G502), 'e70d9670-db2e-11ee-8009-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=Gaming Mouse G502), 'e70d9670-db2e-11ee-800a-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=Gaming Mouse G502), 'e70d9670-db2e-11ee-800b-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=Logitech Extreme 3D), 'e70d9670-db2e-11ee-800c-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=Arcade Stick (Street Fighter IV FightStick TE)), '78a76bc0-2f4d-11ef-8008-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024

(DirectInputDevice(name=GMMK PRO), 'e70dbd80-db2e-11ee-800e-444553540000')
❌ GetProperty error: exception: access violation writing 0x0000000000000024
```

**Comment by benmoran56** [comment]
Yes, the XInput IDs are just the detection order.

**Comment by unexpectedpanda** [comment]
> Yes, the XInput IDs are just the detection order.

Good to know!

> By the way, what kind of application are you building?

A companion app for [MAME](https://mamedev.org). MAME initializes input devices based on the user's API preference, whether that's DirectInput, XInput, raw input, or otherwise. However, the controller order (player 1, player 2, and so on) isn't stable as its based on device enumeration order.

There's a [workaround](https://docs.mamedev.org/advanced/devicemap.html) that lets you set controller order based on device IDs instead of enumeration order, but there's a lot of text editing and futzing about with things like Device Manager in Windows. I'm trying to simplify the process by providing a GUI that works out all the IDs and generates its own config file based on user preferences.

**Comment by caffeinepills** [comment]
Looks like I was possibly just overthinking it. Not sure why it doesn't like the GUID object.
```py
            data = DIPROPGUIDANDPATH()
            data.diph.dwSize = ctypes.sizeof(DIPROPGUIDANDPATH)
            data.diph.dwHeaderSize = ctypes.sizeof(dinput.DIPROPHEADER)
            data.diph.dwObj = 0
            data.diph.dwHow = dinput.DIPH_DEVICE
         
            device.GetProperty(12, ctypes.byref(data.diph))
```
2 of my devices don't implement it, but it atleast works now:
```
DirectInputDevice(name=Razer Naga Trinity) 277e9920-dcba-11ee-8009-444553540000
GUID(745a17a0, 74d3, 11d0, b6, fe, 0, a0, c9, f, 57, da) \\?\hid#vid_1532&pid_0067&mi_01&col02#7&39ff06a5&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
```

**Comment by unexpectedpanda** [comment]
> Looks like I was possibly just overthinking it. Not sure why it doesn't like a Pointer to the GUID object.

Amazing, that did the trick! :) And absolutely saved me from having to do something like learn C# and use SharpDX. Thanks!

Looking forward to playing around with next steps :)
